beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,13 +1733,63 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         result
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block: &Hash256,
+        new_finalized_block: &Hash256,
+    ) -> Result<(), Error> {
+        let old_finalized_slot = self
```

Comment:
> In fact, I think it's safest for us to choose `old_finalized_slot` using `old_finalized_epoch.start_slot(T::EthSpec::slots_per_epoch())` (we can also save a disk read by passing it in from `after_finalization`).

Will do.

> Due to skip slots, it's possible that the block pointed to by `finalized_checkpoint.root` is actually from a higher slot (but not a higher epoch) than `finalized_checkpoint.epoch.start_slot(..)`. If that's the case, we want to prune all the way back to the first slot of the epoch, not the finalized block. (This is kinda complicated, feel free to ask me more about this here/in a DM)

Could you point me at the right place in the spec?  Thanks.

Reply:
Sure! This thing isn't really _explicitly_ said in the spec but I think the [process_slot](https://github.com/ethereum/eth2.0-specs/blob/v0.11.0/specs/phase0/beacon-chain.md#beacon-chain-state-transition-function) function shows (indirectly) that when there's skip slots we update `state.block_roots` with the root of the block from the last non-skipped slot.

Reply:
> Due to skip slots, it's possible that the block pointed to by finalized_checkpoint.root is actually from a higher slot (but not a higher epoch) than finalized_checkpoint.epoch.start_slot(..)

I don't think this is true. I think the slot of the finalized block can be _lower_ than the first slot of the finalized epoch, because when constructing FFG votes, validators include what they perceive the block root at the start of the epoch to be (see: https://github.com/ethereum/eth2.0-specs/blob/v0.11.0/specs/phase0/validator.md#lmd-ghost-vote), and that block can be from any slot `<=` the epoch's start slot. I think this means the current pruning is preferable, as it cleans up any blocks between the finalized block slot and the finalized epoch start slot.

Reply:
Oh you're right @michaelsproul! Sorry, my bad.

Reply:
> I think this means the current pruning is preferable, as it cleans up any blocks between the finalized block slot and the finalized epoch start slot.

This could be a good test case